plugins/out_s3/s3.c

```diff
@@ -967,6 +1098,516 @@ static int cb_s3_init(struct flb_output_instance *ins,
+        ret = -4;
+        goto error;
+    }
+
+    fdout = _open_osfhandle((intptr_t) rh, _O_RDONLY);
+    if (fdout == -1) {
+        ret = -3;
+        goto error;
+    }
+
+    result = WriteFile(wh, body, body_size, &bytes, NULL);
+    if (!result) {
+        ret = -5;
+        goto error;
+    }
+    CloseHandle(wh);
+
+    ret = build_columnify_command(ctx, in_temp_file, out_temp_file, &parquet_cmd);
+    if (ret != 0) {
+        ret = -1;
+        goto error;
+    }
+
+    cmdp = _popen(parquet_cmd, "r");
+    if (cmdp == NULL) {
+        flb_plg_error(ctx->ins, "command %s failed", DEFAULT_PARQUET_COMMAND_EXISTENCE);
+        return -1;
+    }
+    _pclose(cmdp);
+
+    if (fstat(fdout, &stbuf) == -1) {
+        ret = -6;
+        goto error;
+    }
+    parquet_size = stbuf.st_size;
+    parquet_buf = flb_sds_create_size(parquet_size);
+
+    result = ReadFile(rh, parquet_buf, parquet_size, &bytes, NULL);
+    if (!result) {
+        ret = -5;
+        goto error;
+    }
+
+    CloseHandle(rh);
+
+    if (!DeleteFileA((LPTSTR)in_temp_file)) {
+        ret = -6;
+        flb_plg_error(ctx->ins, "DeleteFileA for %s failed", (LPTSTR)in_temp_file);
+    }
+    if (!DeleteFileA((LPTSTR)out_temp_file)) {
+        ret = -6;
+        flb_plg_error(ctx->ins, "DeleteFileA for %s failed", (LPTSTR)out_temp_file);
+    }
+
+    *payload_buf = parquet_buf;
+    *payload_size = parquet_size;
+
+    flb_sds_destroy(parquet_cmd);
+
+    return 0;
+
+error:
+    if (wh != NULL) {
+        CloseHandle(wh);
+    }
+    if (rh != NULL) {
+        CloseHandle(rh);
+    }
+    if (parquet_cmd != NULL) {
+        flb_sds_destroy(parquet_cmd);
+    }
+
+    return ret;
+}
+
+#else
+static const char *get_tmpdir()
+{
+    const char* tmp;
+#ifdef __ANDROID__
+        tmp = "/data/local/tmp";
+#else
+        tmp = "/tmp";
+#endif
+    return tmp;
+}
+
+static int create_tmpfile(char *file_path, char *template, size_t template_len)
+{
+    int ret;
+    int result;
+    flb_sds_t path_buf;
+    const char *tmpdir;
+    size_t tmpdir_len;
+
+    path_buf = flb_sds_create_size(PATH_MAX);
+    if (path_buf == NULL) {
+        goto error;
+    }
+
+    tmpdir = get_tmpdir();
+    tmpdir_len = strlen(tmpdir);
+
+    result = flb_sds_cat_safe(&path_buf, tmpdir, tmpdir_len);
+    if (result < 0) {
+        ret = -1;
+        goto error;
+    }
+
+    result = flb_sds_cat_safe(&path_buf, "/", 1);
+    if (result < 0) {
+        ret = -1;
+        goto error;
+    }
+
+    result = flb_sds_cat_safe(&path_buf, template, template_len);
+    if (result < 0) {
+        ret = -1;
+        goto error;
+    }
+
+    strncpy(file_path, path_buf, flb_sds_len(path_buf));
+    if (mkstemp(file_path) == -1) {
+        flb_errno();
+        ret = -2;
+        goto error;
+    }
+
+    flb_sds_destroy(path_buf);
+
+    return 0;
+
+error:
+    if (path_buf != NULL) {
+        flb_sds_destroy(path_buf);
+    }
+
+    return ret;
+}
+
+static int s3_compress_parquet(struct flb_s3 *ctx,
```

Comment:
multiple definition of same function?

Reply:
Nope. The former one is defined for non-Windows platforms and this one is defined for Windows platform.